src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -214,6 +215,35 @@ private static ReadOnlyMemory<byte> GetCoseAlgorithmFromProtectedHeaders(CoseHea
             return encodedAlg;
         }
 
+        // If we Validate: The caller did specify a COSE Algorithm, we will make sure it matches the specified key and hash algorithm.
+        // If we Add: The caller did not specify a COSE Algorithm, we will add the header for them, rather than throw. protectedHeaders will be altered in this case.
+        private static void ValidateOrAddAlgorithmHeader(CoseHeaderMap protectedHeaders, CoseHeaderMap unprotectedHeaders, KeyType keyType, HashAlgorithmName hashAlgorithm)
+        {
+            int algHeaderValue;
+
+            if (protectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out ReadOnlyMemory<byte> encodedAlg))
+            {
+                int? algHeaderValueNullable = DecodeCoseAlgorithmHeader(encodedAlg);
+                Debug.Assert(algHeaderValueNullable.HasValue, "Algorithm (alg) is a known header and should have been validated in Set[Encoded]Value()");
+                algHeaderValue = algHeaderValueNullable.Value;
+
+                HashAlgorithmName expectedHashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(algHeaderValue, keyType);
+
+                if (expectedHashAlgorithm != hashAlgorithm)
+                {
+                    throw new CryptographicException(SR.Format(SR.Sign1SignHashAlgorithmDoesNotMatchSpecifiedAlg, hashAlgorithm.Name, algHeaderValue));
+                }
+            }
+
+            if (unprotectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out _))
+            {
+                throw new CryptographicException(SR.Sign1SignAlgIsRequired);
+            }
+
+            algHeaderValue = GetCoseAlgorithmHeaderFromKeyTypeAndHashAlgorithm(keyType, hashAlgorithm);
+            protectedHeaders.SetValue(CoseHeaderLabel.Algorithm, algHeaderValue);
```

Comment:
Because this is modifying (and not unmodifying) the user input, you can't do something like

```C#
CoseHeaderMap protectedHeaders = BuildAllMyHeaders();

byte[] ecdsaSigned = CoseSign1Message.Sign(data, protectedHeaders, null, ecdsaKey, HashAlgorithmName.SHA256);
byte[] rsaSigned = CoseSign1Message.Sign(data, protectedHeaders, null, rsaKey, HashAlgorithmName.SHA256);
```

Since the second call will throw that the headers contained an algId that didn't match.

A couple of different approaches:

* Remember that we need to add it, and just slip it in when encoding the header map
  * Basically, add an optional parameter to CoseHeaderMap.Encode for an algId to write down (0 is never an algorithm, so that can represent "don't write this down specially")
* Clone the header map, add the header to the clone
* Add it here, remember that we added it, remove it after encoding.
  * That means the same header map can't be used in two parallel signing operations, which might surprise people.

I'd probably go with the extra parameter to encode.
*

Reply:
Yeah, I thought that this was going to be an issue, will go with first alternative.